### ISSUE TITLE
Some form of flag inheritance for `def` and `extern`

### ISSUE THREAD

**Issue by KAAtheWiseGit** [nushell/nushell#12363]
### Related problem

A common pattern in CLI utilities with subcommands is a set of flags, which are shared between all commands (`--verbose`, `--force`, and so on).

In Nushell, currently, subcommands are independent of their parent commands.  This makes implementing nested commands and `extern` completions for tools like `git` or `apk` tricky, because the global options have to be repeated several times.

### Describe the solution you'd like

My idea is to add some kind of inheritance, so that a custom function or an extern can inherit some other command's flags.  This allows to enable completely separate subcommands (so, `aa x` and `aa y` will, by default, be unrelated), but still provide the option to share flags as an opt-in.

### Describe alternatives you've considered

- Flags as data.  For example, flags could be written down as a `long`, `short`, `doc`, `type`, `default`, `completer` table and be added to a command:
   ```
  const flags = [
  	{
  		long: "verbose",
		short: "v",
		doc: "Print debug information",
  	},
	{
		long: "format",
		doc: "The format for the dates",
		type: string,
		completer: {|| format_func}
	},
	...
  ]

  def command [
  	...flags
  ] {
  	# stuff
  }
  ```

- Automatically inheriting flags for subcommands.  So `def aa x ...` would inherit the flags of `def aa`, with an opt-out flag in `def`.  Not very elegant, and doesn't work when the root command is called `main`.

**Comment by poliorcetics** [comment]
I'm interested in working on this (in my free time so zero timeline) but I wouldn't even know where to start! I'll ask maintainers on discord if they have guidance

**Comment by cablehead** [comment]
From @kubouch on [discord](https://discord.com/channels/601130461678272522/615962413203718156/1271064225703264366):

> This requires a fair bit of design. I've had an idea of being able to define and use standalone signatures, e.g.
```
signature foo_sig = [a: int, b: string]
def foo $foo_sig { ... }
def bar [...$foo_sig, --extra-flag] { ... }
do {|...$foo_sig| ... } 1 'b'
```

**Comment by cablehead** [comment]
I'm not sure if this is possible, but I'd love a convenient way to create commands with this type of usage:

```
command <shared positional args> <sub-command> [--command specific args--]
```

e.g. say a kv command that works with a local embedded store:

```
kv ./store put <key> <value>
kv ./store get <key>
```

**Comment by KAAtheWiseGit** [comment]
Standalone signatures sound like a great idea.  My suggestion is a bit hacky, in that it requires an untyped record which has different structure for the same semantics

**Comment by cablehead** [comment]
Collecting use cases: https://discord.com/channels/601130461678272522/1329144613101109318/1329144613101109318

```
def main [
  --environment (-e): string = "dev"
] {
  // ?
}
def "main sub" [] {
  print $"Env: ($env)"
}
main -e dev sub
-> Env: dev
```

**Comment by ajuckel** [comment]
@cablehead Thanks for copying it over. I will say my idea goes a little further than just argument sharing. In my thinking of it (and the way I've hacked it together with closures) the parent command is responsible for parsing its arguments, and packaging them in to some form of context which is then passed on to the nested command.

```nushell
def main [
  --environment (-e): string = "dev"
  nested?
] {
  print "running main"
  let sub_commands = {
    "my-nested": { my-nested }
  }
  { env: $environment } | do ($sub_commands | get $nested)
}

def "main nested" [] {
  print "main did not run"
}

def my-nested [] {
  let context = $in
  print $"Env: ($context.env)"
}
```

```
$ ./script.nu my-nested
running main
Env: dev
$ ./script.nu -e prod my-nested
running main
Env: prod
```

**Comment by ajuckel** [comment]
I agree with the notion of opt in, but maybe this could be accomplished with just some sigil?

`def "main | nested" []` instead of just `def "main nested" []` to indicate that you want your nested command to run main (and any arguments that are before `nested` on the command line), and pipe its result to your nested function.  Obviously, that makes argument dispatch more complicated, as you can't just look to the first couple tokens to decide which def to run.

I haven't looked at the nushell implementation at all, so I have no idea what sort of effort it would take to accomplish this. Just putting my thoughts out there, for now. For my immediate needs, I just going to continue with my hacky, no-auto-help-supported closure solution.
